### PR TITLE
mgr/rook: persist drive groups

### PR DESCRIPTION
**Pull request by liewegas**

### CODE REVIEW

File: src/pybind/mgr/rook/module.py [hunk 1]
```diff
@@ -523,10 +537,18 @@ def remove_daemons(self, names: List[str]) -> List[str]:
...
+        self._save_drive_groups()
+        return OrchResult(self._apply_drivegroups(specs))
+
+    def _apply_drivegroups(self, ls: List[DriveGroupSpec]) -> List[str]:
+        all_hosts = raise_if_exception(self.get_hosts())
+        result_list: List[str] = []
+        for drive_group in ls:
+            matching_hosts = drive_group.placement.filter_matching_hosts(
+                lambda label=None, as_hostspec=None: all_hosts
+            )
```
Comment (sebastian-philipp) [inline src/pybind/mgr/rook/module.py:551]: I'd like to remove filter_matching_hosts cause the invocation is so awkward. wdyt of`filter_matching_hostssoecs`?

File: src/pybind/mgr/rook/module.py [hunk 2]
```diff
@@ -538,7 +560,23 @@ def apply_drivegroups(self, specs: List[DriveGroupSpec]) -> OrchResult[List[str]
...
+                try:
+                    self._drive_group_map[name] = DriveGroupSpec.from_json(dg)
+                except ValueError as e:
+                    self.log.error(f'Failed to load drive group {name} ({dg}): {e}')
+
+    def _save_drive_groups(self) -> None:
+        json_drive_group_map = {
+            name: dg.to_json() for name, dg in self._drive_group_map.items()
+        }
+        self.set_store("drive_group_map", json.dumps(json_drive_group_map))
```
Comment (sebastian-philipp) [inline src/pybind/mgr/rook/module.py:579]: What I don't like here is that we're hiding those drive groups from Rook by storing them in the MONs. This feels so wrong to me. We've made the same mistake with OCS where managing the Rook CRs from the OCS CRs making is somewhat hard to even enable the Dashboard in an OCS context. And I feel we're making the same mistake all over again here.

Comment (liewegas) [reply src/pybind/mgr/rook/module.py:579]: Rook doesn't understand drive groups--they're a ceph feature.  I don't see how we can not "hide" them from rook.

We could store them in k8s somewhere (ConfigMap?) I guess, but that doesn't actually accomplish anything except make the code more complex.

**Comment by liewegas** [comment]
jenkins test api

**Comment by liewegas** [comment]
https://pulpito.ceph.com/sage-2021-11-04_19:17:56-orch:rook-wip-sage-testing-2021-11-04-1303-distro-basic-smithi/

**Review by sebastian-philipp — APPROVED** [review]

**Pull request merged**